File: day-04/part1.py
from util.run import run
from util.types import Board
from logic.logic import mark_number_if_present, has_winning_row, sum_of_unmarked_numbers


def solution(numbers: list[int], boards: list[Board]) -> int:
    for number in numbers:
        for board in boards:
            mark_number_if_present(number, board)

            if has_winning_row(board):
                return number * sum_of_unmarked_numbers(board)

    raise NoWinningBoardException()


class NoWinningBoardException(Exception):
    pass


if __name__ == "__main__":
    run(solution)

File: day-04/part1.go
package main

import (
	l "github.com/dawidl022/advent-of-code-2021/day-04/logic"
	"github.com/dawidl022/advent-of-code-2021/day-04/util"
)

func solution1(numbers []int, boards []*util.Board) int {
	for _, number := range numbers {
		for _, board := range boards {
			l.MarkNumberIfPresent(number, board)

			if l.HasWinningRow(board) {
				return number * l.SumOfUnmarkedNumbers(board)
			}
		}
	}
	util.Terminate(noWinningBoardError{})
	return -1
}

type noWinningBoardError struct{}

func (e noWinningBoardError) Error() string {
	return "no boards won from given input"
}

func main() {
	util.Run(solution1)
}
